app/src/main/java/com/sample/appconnectsample/ImageCaptureActivity.java

```diff
@@ -0,0 +1,193 @@
+import com.mdsol.babbage.model.DatastoreFactory;
+import com.mdsol.babbage.model.Subject;
+
+import java.io.ByteArrayOutputStream;
+import java.io.FileNotFoundException;
+import java.io.InputStream;
+
+/**
+ * An activity to save an image to AppConnect
+ */
+public class ImageCaptureActivity extends AppCompatActivity {
+
+    public static final String SUBJECT_ID_EXTRA = "appconnectsample.imagecaptureactivity.intent.extra.SUBJECT_ID";
+
+    private static final String TAG = "ImageCaptureActivity";
+
+    private long subjectID;
+    private Bitmap currentImage;
+    private ImageView currentImageView;
+    private Button submitButton;
+    private Button takePictureButton;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.image_capture_activity);
+
+        currentImageView =  (ImageView)findViewById(R.id.currentThumbnail);
+        submitButton = (Button)findViewById(R.id.imageSubmitButton);
+        takePictureButton = (Button)findViewById(R.id.takePictureButton);
+
+        // Get the ID of the subject to save images for
+        Intent intent = getIntent();
+        subjectID = intent.getLongExtra(SUBJECT_ID_EXTRA, 0);
+
+        if (!getApplicationContext().getPackageManager().hasSystemFeature(PackageManager.FEATURE_CAMERA_ANY))
+            takePictureButton.setEnabled(false);
+
+        submitButton.setEnabled(false);
+
+    }
+
+    public void doSelectImageButton(View source) {
+        dispatchSelectImageIntent();
+    }
+
+    public void doTakePictureButton(View source) {
+        dispatchTakePictureIntent();
+    }
+
+    private static final int IMAGE_CAPTURE_INTENT = 1;
+    private static final int SELECT_IMAGE_INTENT = 2;
+
+    private void dispatchTakePictureIntent() {
+        Intent takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
+        if (takePictureIntent.resolveActivity(getPackageManager()) != null) {
+            startActivityForResult(takePictureIntent, IMAGE_CAPTURE_INTENT);
+        }
+    }
+
+    @Override
+    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
+        if (requestCode == IMAGE_CAPTURE_INTENT && resultCode == RESULT_OK) {
+            Bundle extras = data.getExtras();
+            currentImage = (Bitmap) extras.get("data");
+            currentImageView.setImageBitmap(currentImage);
+            submitButton.setEnabled(true);
+        }
+        else if (requestCode == SELECT_IMAGE_INTENT && resultCode == RESULT_OK) {
+            Uri selectedImage = data.getData();
+            try {
+                InputStream imageStream = getContentResolver().openInputStream(selectedImage);
+                currentImage = BitmapFactory.decodeStream(imageStream);
+                currentImageView.setImageBitmap(currentImage);
+                submitButton.setEnabled(true);
+            } catch (FileNotFoundException exception) { }
```

Comment:
This logic could be restructured a bit to avoid duplicating the lines (e.g. look for RESULT_OK in an outer conditional):
```
currentImageView.setImageBitmap(currentImage);
submitButton.setEnabled(true);
```